### ISSUE TITLE
Raw Content

### ISSUE THREAD

**Issue by prabhunatarajan14** [Kiran143604/Hackathonscripts#2]
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# -------------------------------
# AZURE OPENAI SETUP
# -------------------------------
USE_AI = True  # turn to False if you want to disable AI reasoning

from openai import AzureOpenAI

client = None
try:
    client = AzureOpenAI(
        api_key="YOUR_KEY_HERE",   # <--- PUT YOUR KEY HERE
        api_version="2024-02-01",
        azure_endpoint="https://mobpochack25-openai-18.openai.azure.com/"
    )
except Exception as e:
    print("⚠️ Azure client initialization failed:", e)
    USE_AI = False

# -------------------------------
# AI Reasoning Function
# -------------------------------
def ask_ai_reasoning(input_data, predicted_label):
    if not USE_AI or client is None:
        return "AI reasoning disabled."

    try:
        prompt = f"""
        You are an AI assistant helping with store audit reassignment decisions.
        Explain the prediction briefly.

        Store / Audit Input:
        {input_data}

        ML Model Prediction: {predicted_label}
        """

        response = client.chat.completions.create(
            model="mobpochack25-openai-18",  # deployment name
            messages=[
                {"role": "system", "content": "You explain audit reassignment reasoning simply."},
                {"role": "user", "content": prompt},
            ],
            max_tokens=150
        )

        return response.choices[0].message.content

    except Exception as e:
        return f"AI Error: {str(e)}"

# -------------------------------
# Sample synthetic training data
# -------------------------------
import numpy as np

n_samples = 500
data = {
    'auditor_location_distance': np.random.randint(1, 60, n_samples),
    'auditor_availability': np.random.randint(0, 2, n_samples),
    'auditor_skill_level': np.random.randint(1, 4, n_samples),
    'store_risk_score': np.random.randint(1, 10, n_samples),
    'audit_priority': np.random.randint(1, 4, n_samples),
    'traffic_condition': np.random.randint(0, 2, n_samples),
    'weather_condition': np.random.randint(0, 2, n_samples),
    'audit_deadline_hours': np.random.randint(1, 10, n_samples),
    'store_open_status': np.random.randint(0, 2, n_samples)
}

# -------------------------------
# Create target variable
# -------------------------------
target = []
for i in range(n_samples):
    score = 0
    if data['auditor_availability'][i] == 0: score += 2
    if data['auditor_location_distance'][i] > 30: score += 1
    if data['store_risk_score'][i] > 7: score += 1
    if data['audit_priority'][i] > 3: score += 1
    if data['traffic_condition'][i] == 1 or data['weather_condition'][i] == 1: score += 1
    if data['audit_deadline_hours'][i] < 4: score += 1
    if data['store_open_status'][i] == 0: score += 2
    target.append(1 if score >= 3 else 0)

df = pd.DataFrame(data)
df['reassignment_needed'] = target

# -------------------------------
# Train model
# -------------------------------
X = df.drop('reassignment_needed', axis=1)
y = df['reassignment_needed']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

print("\n🚀 MODEL TRAINED")
print(f"Accuracy: {accuracy_score(y_test, model.predict(X_test)):.2f}")
print("Confusion Matrix:\n", confusion_matrix(y_test, model.predict(X_test)))
print("Classification Report:\n", classification_report(y_test, model.predict(X_test)))

# -------------------------------
# Example prediction
# -------------------------------
example = pd.DataFrame({
    'auditor_location_distance': [35],
    'auditor_availability': [0],
    'auditor_skill_level': [2],
    'store_risk_score': [9],
    'audit_priority': [4],
    'traffic_condition': [1],
    'weather_condition': [0],
    'audit_deadline_hours': [3],
    'store_open_status': [0]
})

prediction = model.predict(example)[0]
print("\n🔍 ML Prediction:", "Reassign" if prediction == 1 else "No Reassign")

# AI explanation
ai_reason = ask_ai_reasoning(example.to_dict(), "Reassign" if prediction == 1 else "No Reassign")
print("🤖 AI Explanation:", ai_reason)

# -------------------------------
# Reassignment logic
# -------------------------------
shops = [
    {"id": 1, "name": "Shop A", "status": "closed"},
    {"id": 2, "name": "Shop B", "status": "open"},
    {"id": 3, "name": "Shop C", "status": "open"},
]

auditor = {"name": "John", "assigned_shop": "Shop A"}

if prediction == 1 or auditor["assigned_shop"] == "Shop A":
    print("\n⚠️ Reassignment triggered...")
    new_shop = next((shop for shop in shops if shop["status"] == "open"), None)
    if new_shop:
        auditor["assigned_shop"] = new_shop["name"]
        print(f"✅ Auditor reassigned to {new_shop['name']}")
    else:
        print("❌ No open shop available.")
else:
    print("\n👍 No reassignment needed.")

# -------------------------------
# USER INPUT SECTION
# -------------------------------
print("\n🔽 Enter details for custom prediction:")

auditor_location_distance = int(input("Distance (km): "))
auditor_availability = int(input("Availability (0=busy, 1=free): "))
auditor_skill_level = int(input("Skill level (1-4): "))
store_risk_score = int(input("Risk score (1-10): "))
audit_priority = int(input("Priority (1-4): "))
traffic_condition = int(input("Traffic (0=normal, 1=heavy): "))
weather_condition = int(input("Weather (0=clear, 1=bad): "))
audit_deadline_hours = int(input("Hours remaining: "))
store_open_status = int(input("Store open (0/1): "))

user_input = pd.DataFrame({
    'auditor_location_distance': [auditor_location_distance],
    'auditor_availability': [auditor_availability],
    'auditor_skill_level': [auditor_skill_level],
    'store_risk_score': [store_risk_score],
    'audit_priority': [audit_priority],
    'traffic_condition': [traffic_condition],
    'weather_condition': [weather_condition],
    'audit_deadline_hours': [audit_deadline_hours],
    'store_open_status': [store_open_status],
})

prediction = model.predict(user_input)[0]
out = "Reassign" if prediction == 1 else "No Reassign"

print("\n🔍 RESULT:", out)

print("🤖 AI Explanation:", ask_ai_reasoning(user_input.to_dict(), out))
